sygnal/notifications.py

```diff
@@ -14,32 +14,33 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-import typing
-from typing import Any, Dict, List, Optional
 
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Union
+
+from opentracing import Span
 from prometheus_client import Counter
 
 from .exceptions import InvalidNotificationException, NotificationDispatchException
 
-if typing.TYPE_CHECKING:
+if TYPE_CHECKING:
     from .sygnal import Sygnal
 
 
 class Tweaks:
-    def __init__(self, raw):
-        self.sound = None
+    def __init__(self, raw) -> None:
+        self.sound: Optional[str] = None
 
         if "sound" in raw:
             self.sound = raw["sound"]
 
 
 class Device:
     def __init__(self, raw):
-        self.app_id = None
-        self.pushkey = None
-        self.pushkey_ts = 0
-        self.data = None
-        self.tweaks = None
+        self.app_id: Optional[str] = None
+        self.pushkey: Optional[str] = None
```

Comment:
These two are kind of fake optional.

Reckon you could do the dance to make it so that they're not annotated as optional? It probably needs to be something like:

```
// after if "app_id" not in raw: .... raise ...
raw_app_id = raw["app_id"]
if not isinstance(raw_app_id, str):
    raise InvalidNotificationException("Device with non-string app_id")
self.app_id: str = raw_app_id
```

(please do similarly for the pushkey, if you don't mind)

This just makes your life easier around the rest of the code that otherwise has to deal with this hypothetical (but impossible) case that these fields are `None`

Reply:
We can simplify these constructors for `Tweaks`, `Device` and `Counts` by using `dict.get`:
```py
    def __init__(self, raw: Dict[str, Any]):
         if "app_id" not in raw:
             raise InvalidNotificationException("Device with no app_id")
         if "pushkey" not in raw:
             raise InvalidNotificationException("Device with no pushkey")

       self.app_id: str = raw["app_id"]
       self.pushkey: str = raw["pushkey"]
       self.pushkey_ts: int = raw.get("pushkey_ts", 0)
       self.data: Optional[Dict[str, Any]] = raw.get("data")
       self.tweaks = Tweaks(raw.get("tweaks", {}))
```
(or similar)

(nb: `tweaks` is not `Optional`)

and if we wanted validation, we could define a helper function similar to my suggestion for `get_config`

Reply:
We definitely want validation.

```
         if "app_id" not in raw:
             raise InvalidNotificationException("Device with no app_id")
       self.app_id: str = raw["app_id"]
```
is a worst-of-both worlds approach where you get neither validation nor type errors (since `Any` will unify with `str` and everything pretends its fine)

Reply:
I added a helper method to do some validation, which is quite similar to the `get_config` helper function, which makes me wonder if it makes sense to combine them? It seems that you would lose the specificity of the error message in `get_config`, but I don't know how to weigh that against having somewhat repeated code in a file. Thoughts?